### DISCUSSION TITLE
Multi-select breaks first normal form

### DISCUSSION THREAD

**Discussion by userbal** [pocketbase/pocketbase#1489] [Q&A] [answered]
I don't know if this is a real issue or something that can be fixed so I thought I'd bring it up as a discussion. Currently if you create a select type in the ui and save multiple values it saves an array of those values in the database. My understanding is that this breaks first normal form and can slow down queries and joins. Does Pocketbase have workaround for this? Would it speed things up to create multiple rows or tables for select types?

Thanks for sharing this awesome project!

**Comment by ganigeorgiev** [comment, marked as answer]
All multi-value fields store their data (eg. the references) as json array and this is fine.

The concept for most of the DNFs were created long before modern SQL features and types were added and you shouldn't follow them strictly for all cases. Most databases (including SQLite) nowadays have integrated [JSON functions and operators](https://www.sqlite.org/json1.html) that allow querying and sorting JSON data more efficiently.

> Would it speed things up to create multiple rows or tables for select types?

It depends. Making multiple large queries or joins are not always faster especially if you are joining a large dataset and have aggregations. The main eventual benefits by keeping your data "normalized" (eg. in a junction tables) would be that:
- you can make use of the db cascade operations and the foreign key constraints (in PocketBase this is all handled on application level, so it is not a concern)
- you can define more easily an index for each individual value, but there is no guarantee that the query planner will choose your index(es), so it kind of depends on the query. Additionally, at least for SQLite, there is support for partial/expression indexes which can be used for indexing the `json_extract` value(s) (or even create separate generated/computed columns). PocketBase currently doesn't have a way to define indexes from the Admin UI but this will be handled in the near future as part of #345 and #544.

**Reply by denisgurev** [reply]
I'm using heavily the multiple relations feature (some of my records have more than 50000 ids associated to them) and the performance is great so far even without indexes, but the ui I think could be improved. It is slow when editing records with a lot of relations and sometimes it is very hard to find the item you want to select because the dropdown doesn't show everything and you have to constantly click on the "load more" button. A searchbar like in the main listing would be nice if it could be added.

**Reply by ganigeorgiev** [reply]
@denisgurev I'm aware of these drawbacks. The relations picker, and the relations visualization in general, will be improved in the next v0.12/v0.13 releases. You can subscribe/watch https://github.com/pocketbase/pocketbase/issues/976.
